127 Add support for more types in memory footprinting mode

cz4rs opened the pull request:
fixes #127

cz4rs committed:
#127 implement memory footprinting for void*

cz4rs committed:
#127 implement memory footprinting for std::stack

cz4rs committed:
#127 improve std::map and std::set support

Move reconstructed elements when deserializing map-like containers to
avoid calling copy constructor for std::unique_ptr.

lifflander reviewed (approved):
This is looking good to me. We definitely should make sure that all the serializers work with `std::unique_ptr` as much as possible. That might have been overlooked since support for normal serialization of `std::unique_ptr` was implemented fairly recently.

lifflander commented:
I'm not going to merge this because I'm not sure if you are done yet.

cz4rs committed:
#127 add Footprinter to serializer variadic args

cz4rs committed:
#127 fix virtual serialization for std::shared_ptr and raw pointers

- enable raw pointer and std::shared_ptr serialization for all serializers
- modify virtual serialization tests to cover raw and shared pointers

cz4rs commented:
@lifflander 
> I'm not going to merge this because I'm not sure if you are done yet.

there is a bit of work to support remaining required types, but the bigger challenge that I'm facing right now is with virtual serialization
- I have started to implement virtual serialization for `LoadModel` on vt side, based on examples and comments in `dispatch_virtual.h`
- I've had to take a step back with e5fe0f2 - if shared_ptr's `serialize` is available only to `Footprinter`, I get:
```
[build] ../lib/checkpoint/src/checkpoint/container/shared_ptr_serialize.h: In instantiation of ‘void checkpoint::serialize(SerializerT&, std::shared_ptr<_Tp>&) [with SerializerT = checkpoint::PackerBuffer<checkpoint::buffer::ManagedBuffer>; T = vt::vrt::collection::balance::LoadModel]’:
[build] ../lib/checkpoint/src/checkpoint/dispatch/dispatch_serializer_nonbyte.h:86:14:   required from ‘static void checkpoint::dispatch::BasicDispatcher<SerializerT, T>::serializeNonIntrusive(SerializerT&, T&) [with SerializerT = checkpoint::PackerBuffer<checkpoint::buffer::ManagedBuffer>; T = std::shared_ptr<vt::vrt::collection::balance::LoadModel>]’
(...)
[build] ../lib/checkpoint/src/checkpoint/dispatch/vrt/serialize_instantiator.h:91:80:   required from ‘void checkpoint::instantiateObjSerializer() [with ObjT = vt::vrt::collection::balance::PerCollection; SerializerTs = {checkpoint::Footprinter, checkpoint::PackerBuffer<checkpoint::buffer::ManagedBuffer>, checkpoint::PackerBuffer<checkpoint::buffer::UserBuffer>, checkpoint::PackerBuffer<checkpoint::buffer::IOBuffer>, checkpoint::UnpackerBuffer<checkpoint::buffer::UserBuffer>, checkpoint::UnpackerBuffer<checkpoint::buffer::IOBuffer>, checkpoint::Sizer}]’
[build] ../src/vt/vrt/collection/balance/model/per_collection.h:69:3:   required from here
[build] ../lib/checkpoint/src/checkpoint/container/shared_ptr_serialize.h:54:21: error: no matching function for call to ‘serializeSharedPtr(checkpoint::PackerBuffer<checkpoint::buffer::ManagedBuffer>&, std::shared_ptr<vt::vrt::collection::balance::LoadModel>&)’
[build]    54 |   serializeSharedPtr(s, ptr);
```
(as you can see, `checkpoint::PackerBuffer` is used along the way)
- the other issue that I haven't figured out yet on `vt` side (https://github.com/DARMA-tasking/vt/pull/1013/commits/98360e8aa6618b6fdc3ed46845c532b0f861a697)
```
[build] ../lib/checkpoint/src/checkpoint/dispatch/dispatch_serializer_nonbyte.h: In instantiation of ‘void checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::apply(SerializerT&, T*, checkpoint::SerialSizeType, checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::hasNotVirtualSerialize<U>*) [with U = vt::vrt::collection::balance::LoadModel; SerializerT = checkpoint::Footprinter; T = vt::vrt::collection::balance::LoadModel; Dispatcher = checkpoint::dispatch::BasicDispatcher<checkpoint::Footprinter, vt::vrt::collection::balance::LoadModel>; checkpoint::SerialSizeType = long unsigned int; checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::hasNotVirtualSerialize<U> = vt::vrt::collection::balance::LoadModel]’:
(...)
[build] ../src/vt/vrt/collection/balance/lb_invoke/lb_manager.h:265:7:   required from ‘void vt::vrt::collection::balance::LBManager::serialize(Serializer&) [with Serializer = checkpoint::Footprinter]’
(...)
[build] ../lib/checkpoint/src/checkpoint/checkpoint_api.impl.h:97:50:   required from ‘std::size_t checkpoint::getMemoryFootprint(T&) [with T = vt::vrt::collection::balance::LBManager; std::size_t = long unsigned int]’
[build] ../tests/unit/runtime/test_memory_footprint.cc:83:48:   required from ‘void vt::tests::unit::printMemoryFootprint(T*) [with T = vt::vrt::collection::balance::LBManager]’
[build] ../tests/unit/runtime/test_memory_footprint.cc:96:38:   required from here
[build] ../lib/checkpoint/src/checkpoint/dispatch/dispatch_serializer_nonbyte.h:218:5: error: no matching function for call to ‘checkpoint::dispatch::SerializerDispatchNonByte<checkpoint::Footprinter, vt::vrt::collection::balance::LoadModel, checkpoint::dispatch::BasicDispatcher<checkpoint::Footprinter, vt::vrt::collection::balance::LoadModel> >::applyStatic(checkpoint::Footprinter&, vt::vrt::collection::balance::LoadModel*&, checkpoint::SerialSizeType&)’
[build]   218 |     applyStatic(s, val, num);
[build]       |     ^~~~~~~~~~~
[build] ../lib/checkpoint/src/checkpoint/dispatch/dispatch_serializer_nonbyte.h:163:8: note: candidate: ‘template<class U> void checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::applyStatic(SerializerT&, T*, checkpoint::SerialSizeType, checkpoint::dispatch::SerializerDispatchNonByte<SerializerT, T, Dispatcher>::hasInSerialize<U>*) [with U = U; SerializerT = checkpoint::Footprinter; T = vt::vrt::collection::balance::LoadModel; Dispatcher = checkpoint::dispatch::BasicDispatcher<checkpoint::Footprinter, vt::vrt::collection::balance::LoadModel>]’
[build]   163 |   void applyStatic(
[build]       |        ^~~~~~~~~~~
(...)
[build] ../lib/checkpoint/src/checkpoint/dispatch/dispatch_serializer_nonbyte.h:114:11: error: no type named ‘type’ in ‘struct std::enable_if<false, vt::vrt::collection::balance::LoadModel>’
[build]   114 |     using hasInSerialize =
```
the intrusive `serialize` method in `LoadModel` is not detected for some reason - do you have any idea why? as far as I can see the implementation seems correct

cz4rs commented:
if you prefer clang output, build log is available (for example) here: https://github.com/DARMA-tasking/vt/pull/1013/checks?check_run_id=1162292999

lifflander commented:
> * I've had to take a step back with [e5fe0f2](https://github.com/DARMA-tasking/checkpoint/commit/e5fe0f2e4b11694dff092a9b0b3c56c30dea9881) - if shared_ptr's `serialize` is available only to `Footprinter`, I get:
> (as you can see, `checkpoint::PackerBuffer` is used along the way)

For virtual serialize to use, you need to define your serializer as part of a registration macro. Add to this list:

https://github.com/DARMA-tasking/checkpoint/blob/7c8115e7d63eef9cc8dcaaf8f2c9c5fd31089cae/src/checkpoint/serializers/serializers_headers.h#L55-L61

However, I think it's a bug that it doesn't tell you it wasn't registered properly. It should check that when you use virtual serialize. But, once you add it to that list, it should be dispatched properly.

> * the other issue that I haven't figured out yet on `vt` side ([DARMA-tasking/vt@98360e8](https://github.com/DARMA-tasking/vt/commit/98360e8aa6618b6fdc3ed46845c532b0f861a697))
> the intrusive `serialize` method in `LoadModel` is not detected for some reason - do you have any idea why? as far as I can see the implementation seems correct

I'm looking into this now. Not sure off hand.

lifflander commented:
(I'm making this a draft so it doesn't get merged accidentally before it is complete)

cz4rs commented:
> For virtual serialize to work properly, you need to define your serializer as part of a registration macro. Add to this list:

Yep, `Footprinter` serializer was missing on that list, but the logs were produced after I have added it with efe836cd842645c5db95cfd0d30a29708b7c2416


> (I'm making this a draft so it doesn't get merged accidentally before it is complete)

Oops, I forgot to mark it as draft myself - sorry about that!

cz4rs committed:
#127 add an arbitrary value to memory footprint

cz4rs commented:
@PhilMiller (following from the other issue)
> In the scheduler or active messenger, newly-enqueued messages should have their size added to some gauge variable to track their total footprint, and be subtracted when they're processed. That variable's value can then be added to the scheduler component's footprint when that's being calculated. If there's no method on the footprinter to add an arbitrary value to the computed footprint, that should get implemented.

can it be as simple as d2a8c37?

lifflander commented:
> > For virtual serialize to work properly, you need to define your serializer as part of a registration macro. Add to this list:
> 
> Yep, `Footprinter` serializer was missing on that list, but the logs were produced after I have added it with [efe836c](https://github.com/DARMA-tasking/checkpoint/commit/efe836cd842645c5db95cfd0d30a29708b7c2416)
> 
> > (I'm making this a draft so it doesn't get merged accidentally before it is complete)
> 
> Oops, I forgot to mark it as a draft myself - sorry about that!

Sorry, I missed that it was part of the list. Probably had an old version of the branch.

lifflander commented:
> @PhilMiller (following from the other issue)
> 
> > In the scheduler or active messenger, newly-enqueued messages should have their size added to some gauge variable to track their total footprint, and be subtracted when they're processed. That variable's value can then be added to the scheduler component's footprint when that's being calculated. If there's no method on the footprinter to add an arbitrary value to the computed footprint, that should get implemented.
> 
> can it be as simple as [d2a8c37](https://github.com/DARMA-tasking/checkpoint/commit/d2a8c37a0ec20941fcb36443e1d4ca3a1b4d16f8)?

That looks reasonable to me if we only want to add it to the total sum, but we might want something that can be called in `serialize` recursively on the footprinter `SerializerT`.

lifflander commented:
> > For virtual serialize to work properly, you need to define your serializer as part of a registration macro. Add to this list:
> 
> Yep, `Footprinter` serializer was missing on that list, but the logs were produced after I have added it with [efe836c](https://github.com/DARMA-tasking/checkpoint/commit/efe836cd842645c5db95cfd0d30a29708b7c2416)
> 
> > (I'm making this a draft so it doesn't get merged accidentally before it is complete)
> 
> Oops, I forgot to mark it as a draft myself - sorry about that!

I've taken a look and don't see anything obviously wrong. Clang seems to get the Footprinter right at least (not `PackerBuffer`) if I'm reading it correctly.

I think the dispatch traits might not be triggering for some reason. Do you have an overload for shared_ptr that calls into the virtual serialize when footprinting?

lifflander commented:
Checkpoint by default does not provide a serialize for shared_ptr on purpose. This is because it often has shared ownership so naively serializing it can end up with multiple copies and broken references during de-serialization.

cz4rs commented:
> Checkpoint by default does not provide a serialize for shared_ptr on purpose. This is because it often has shared ownership so naively serializing it can end up with multiple copies and broken references during de-serialization.

so I understand that keeping the `enable_if` is desirable (to make it work only with Footprinter)

lifflander commented:
> > Checkpoint by default does not provide a serialize for shared_ptr on purpose. This is because it often has shared ownership so naively serializing it can end up with multiple copies and broken references during de-serialization.
> 
> so I understand that keeping the `enable_if` is desirable (to make it work only with Footprinter)

Yes, exactly.

cz4rs commented:
e0124ba adds a testcase displaying the problems with virtual serialization and `shared_ptr`, to isolate it / make it easierr to analyze
I am currently analyzing `serialize_instantiator.h` (and the rest of the `vrt`) to see why `PackerBuffer` is picked from `SerializerTs`, where we have all the serializers present
```
../src/checkpoint/dispatch/vrt/serialize_instantiator.h:91:80:   required from ‘void checkpoint::instantiateObjSerializer() [with ObjT = checkpoint::tests::unit::test_3::TestDerived2; SerializerTs = {checkpoint::Footprinter, checkpoint::PackerBuffer<checkpoint::buffer::ManagedBuffer>, checkpoint::PackerBuffer<checkpoint::buffer::UserBuffer>, checkpoint::PackerBuffer<checkpoint::buffer::IOBuffer>, checkpoint::UnpackerBuffer<checkpoint::buffer::UserBuffer>, checkpoint::UnpackerBuffer<checkpoint::buffer::IOBuffer>, checkpoint::Sizer}]’
../tests/unit/test_virtual_serialize.cc:503:3:   required from here
../src/checkpoint/container/shared_ptr_serialize.h:54:21: error: no matching function for call to ‘serializeSharedPtr(checkpoint::PackerBuffer<checkpoint::buffer::ManagedBuffer>&, std::shared_ptr<int>&)’
```

cz4rs committed:
#127 improve std::list and std::deque support

Move reconstructed elements when deserializing lists and deques to avoid
calling copy constructor for std::unique_ptr.

lifflander commented:
The function `allocateConstructForPointer` is supposed to call `SerializeAsVirtualIfNeeded` which checks to see if virtual serialization is needed or not. It looks like this might not be working properly?

cz4rs committed:
#123 implement memory footprinting for std::list<T>::iterator

cz4rs committed:
#123 support custom deleters when serializing std::unique_ptr

cz4rs commented:
to do:
- [ ] fix virtual serialization / bring back `enable_if`s to check that we use `Footprinter`
- [ ] change `std::_List_iterator<T>` to `std::list<T>::iterator`

cz4rs commented:
@lifflander 
virtual serialization (`std::shared_ptr`) commit: 8157461 
uncommenting the `enable_if` in `shared_ptr_serialize.h` triggers the bug, resulting in errors during compilation

testcase: `test_virtual_serialize.cc `

cz4rs committed:
#127 separate test for virtual serialize / shared_ptr

cz4rs commented:
@lifflander 
for the virtual serialization / shared_ptr issue, please see f5f1782

lifflander commented:
> @lifflander
> for the virtual serialization / shared_ptr issue, please see [f5f1782](https://github.com/DARMA-tasking/checkpoint/commit/f5f178204c168dd3d839db9ee901843e8dee8b77)

@cz4rs Ok, I've solved one problem. Right now, the way virtual serialize works is that it instantiates for all serializer types because it is dynamically dispatched. That is why you are getting the error with Clang when the enable_if is present.

lifflander commented:
The other problem is that `checkpoint::SerializableTraits<T>` checks for `checkpoint::Serializer` to be dispatched on a intrusive or non-intrusive serialize function. However, when enable_if is on the `checkpoint::Footprinter` it doesn't check for convertibility from `checkpoint::Footprinter` to `checkpoint::Serializer`.

lifflander committed:
#127: traits: template on SerializerT (does not check convertibility)

lifflander committed:
#127: virtual: check serializability with SerializerT before instantiating

lifflander committed:
#127: container: add enable_if at first call to check dispatch directly

lifflander committed:
#127: test: push up to top-level class so filter applies for shared_ptr<int>

lifflander commented:
I think I fixed all the bugs (except for list iterator, which I'll leave for you).

- I expanded `SerializableTraits<T>` to `SerializableTraits<T,U>` where `U` is a `SerializerT` so it doesn't assume `checkpoint::Serializer` which leads to convertibility issues.
- For the automatic instantiation in virtual serialize, I first check that the object is actually serializable with the `SerializerT` before trying to instantiate it with the pack of `SerializerTs`
- I put the `std::enable_if` at the top-level of `serialize(SerializerT& s, std::shared_ptr<T>& ptr)` so it can filter out early
- Then, to make the final test compile, I had to also push the enable_if check up into the serializer for `TestDerived2`. The problem is that if `TestDerived2` is serializable with any serializer it assumes that its children will be (I'm still thinking about if there is a way to recursively check that at compile-time). For now, the way to avoid that is to make it's children be serializable or push the check up (which I decided to do in the code).

cz4rs committed:
#127: ci: enable building examples

- enable building examples in Dockerfiles
- use LABEL instead of deprecated MAINTAINER instruction
- fix serializability traits test (checkpoint_example_3.cc)

cz4rs committed:
#127: container: filter serializers early for raw pointers

cz4rs committed:
#127: test: extend virtual serialization test for Footprinter

cz4rs committed:
#127: traits: add default template parameter for SerializableTraits

cz4rs committed:
#127 remove serializer for std::list iterator

cz4rs committed:
#127 do not reconstruct elements when in Footprinting mode

cz4rs commented:
@PhilMiller can you please have a look at this? another approval would be good since Jonathan has also added some commits here

cz4rs committed:
#127 implement memory footprinting for std::atomic

cz4rs committed:
#127 implement memory footprinting for std::thread

PhilMiller commented on src/checkpoint/container/raw_ptr_serialize.h:
```diff
@@ -81,29 +84,23 @@ void serializeRawPtr(SerializerT& s, T* ptr) {
+  typename SerializerT
 >
-void serializeFilePtr(SerializerT& s, FILE* ptr) {
+void serializeRawPtr(SerializerT& s, FILE* ptr) {
```
In the future, if we want to be very exacting about this, or need to support a broader range of possibly-incomplete types, we could use the `type_if_defined` template described [here](https://devblogs.microsoft.com/oldnewthing/20190712-00/?p=102690)

PhilMiller commented on src/checkpoint/container/vector_serialize.h:
```diff
@@ -52,13 +52,23 @@
+
 template <typename Serializer, typename T, typename VectorAllocator>
 void serialize(Serializer& s, std::vector<T, VectorAllocator>& vec) {
   if (s.isFootprinting()) {
```
Since `serializeVectorMeta` is now polymorphic on footprinting or not, we could simplify the logic here a bit, to

```
serializeVectorMeta(s, vec);
size_t num_elements = s.isFootprinting() ? vec.capacity() : vec..size();
dispatch::serializeArray(s, &vec[0], num_elements);
```

And have the `s.countBytes(vec)` in the meta function.

PhilMiller commented on src/checkpoint/serializers/footprinter.h:
```diff
@@ -61,7 +61,7 @@ struct Footprinter : Serializer {
   }
 
   template<typename T>
```
While we're working here, could you also add a simple `void addBytes(size_t s)` method that can be used for containers in the runtime that might keep track of the footprint of objects they're holding, but which can't be traversed directly (e.g. the message queues)

PhilMiller replied:
Body would literally be `num_bytes_ += s;`

PhilMiller commented:
Get the `Footprinter::addBytes` in, and I'm +1 on this.

cz4rs committed:
#127 allow adding arbitrary number of bytes to memory footprint

cz4rs committed:
#127 simplify std::vector serialization

cz4rs commented:
> Get the `Footprinter::addBytes` in, and I'm +1 on this.

`addBytes` added and vector serialization simplified :ok_hand:

PhilMiller commented on tests/unit/test_footprinter.cc:
```diff
@@ -98,6 +113,15 @@ TEST_F(TestFootprinter, test_fundamental_types) {
   }
 
+  {
+    int i = 3;
```
@cz4rs I just found that this generates a warning:

```
/Users/pbmille/repos/vt/lib/checkpoint/tests/unit/test_footprinter.cc:117:9: warning: declaration shadows a local variable [-Wshadow]
    int i = 3;
        ^
/Users/pbmille/repos/vt/lib/checkpoint/tests/unit/test_footprinter.cc:80:7: note: previous declaration is here
  int i;
      ^
```

Please open an issue and address it, by using a different name or something

cz4rs replied:
#134 created
